### DISCUSSION TITLE
How to request changes to Symfony Expression language

### DISCUSSION THREAD

**Discussion by tmbenhura** [symfony/symfony#48463] [Q&A]
I was wondering how to request changes to the Expression Language engine.

There are 3 specific changes, which may be unique to my particular use case:
1) Unboxing: The ability to unbox object operands implementing a method, for example, __toValue or supporting a specific interface. An alternative to unboxing would be to allow objects to implement an interface providing operator methods, e.g...

```
<?php

interface Operand
{
    public function or(mixed $right): bool;
    public function and(mixed $right): bool;
    public function not(mixed $right): bool;
    public function eq(mixed $right): bool;
    public function lt(mixed $right): bool;
    public function lte(mixed $right): bool;
    public function gte(mixed $right): bool;
    public function add(mixed $right): self;
    public function sub(mixed $right): self;
    public function mul(mixed $right): self;
    public function div(mixed $right): self;
    public function mod(mixed $right): self;
}  
```

2) Overloading the + operator to concatenate strings.

3) Returning null when a non-string operand is null.

2 and 3 are possibly breaking changes to the engine.

In my particular use case, I am trying to create a Groovy-like emulation layer.

I have implemented this in a fork. I was also still figuring out how the package tests are run, so sadly no tests have been written.

https://github.com/tmbenhura/expression-language/commits/6.1/Node

**Comment by stof** [comment]
As one of the goal of the component is to be able to compile the expression into a PHP expression with no dedicated runtime, this is not doable AFAICT.
Your branch is implementing this only for the evaluation mode.

**Reply by tmbenhura** [reply]
Hi @stof , thank you for your reply.

I have managed to get it working with the changes as implemented in the fork. But, I think it would be beneficial to have the changes be in the mainstream project.

In my use case, I have PHP classes that implement JavaInt, JavaString and JavaDecimal (and so on) behaviours. The problem is, without unboxing, you cannot add a JavaInt + JavaInt as the expression engine does not allow objects to be operands. You can only invoke methods on the objects.

Yes, my branch only implements unboxing at evaluation and does not box the resulting native int, or native string, back to a JavaInt or JavaString. Though having the engine support operands with an interface similar to the example given, would allow for this. I was looking to see how best to implement this before committing to a particular implementation.

**Reply by tmbenhura** [reply]
In the case of an Operand interface, I think it adds more flexibility to the engine. Suppose you have a Money type, you could pass the objects to an expression to calculate Tax, for example, receiving a Money type as the result, without having to hack the package or create a custom fork.

**Reply by stof** [reply]
As I said, you did not managed to make it work in the fork. You only made it work for the evaluation mode (which is the easy one of the 2 modes). We cannot accept a feature if it does not work for both modes.

**Reply by tmbenhura** [reply]
My apologies, by 2 modes I thought you meant boxing and unboxing. I just realised you meant compilation and evaluation.
